### ISSUE TITLE
Bug: reflector.go stores objects in cache by name only, broken in namespaces

### ISSUE THREAD

**Issue by derekwaynecarr** [GoogleCloudPlatform/kubernetes#2294]
This is broken for resources that are namespaced.

We need to store objects in the cache by namespace+name.

Example:
Two resources with same name in different namespaces will cause a cache clash.

**Comment by ironcladlou** [comment]
This is now a major issue in OpenShift where we use Reflectors with ListWatcher implementations that span namespaces (e.g. watch all builds or deployments in order to advance states). Any name collisions on these resources will result in all sorts of wackiness in FIFOs (due to event compression) and caches (due to last-in-wins overwriting).

Is anybody working on the UID fix? If not, we can take a stab at a patch.

**Comment by derekwaynecarr** [comment]
@ironcladlou - all things have UIDs now, please feel free to submit a fix.

**Comment by smarterclayton** [comment]
Do we need a version of the reflector that caches by namespace+name?  If so, we may want to extract "uniqueness" into its own function and make that be something the reflector can be parameterized with.

----- Original Message -----

**Comment by deads2k** [comment]
I'm not sure the UID fixes the problem for us.  cache.Store has a .Get(name) method.  Some references are based on name and namespace (anything with this name regardless of what its UID is), so we need a logical .Get(namespace, name), .Get(UID) won't do what we want to express.

**Comment by pmorie** [comment]
@deads2k @smarterclayton I have been toying with the idea of separating the key used by the reflector with the interface that a client of a store has.  The reflector needs only a single globally unique value to address objects by (ex: uid), while the clients of the store the reflector is writing into may need other interfaces (ex: namespace/name).  I haven't thought through all the implications of differentiating the interfaces, just recording here before I forget.

**Comment by ironcladlou** [comment]
@deads2k 
>I'm not sure the UID fixes the problem for us. cache.Store has a .Get(name) method. Some references are based on name and namespace (anything with this name regardless of what its UID is), so we need a logical .Get(namespace, name), .Get(UID) won't do what we want to express.

Good point. UID doesn't solve it for the client interface.

**Comment by smarterclayton** [comment]
You don't have to require the reflector to use uid - you just need the person who inits the reflector and the person who calls the reflector agree on what unique is.  I can imagine reflectors that would unique on attributes on the object that is not UID or name+namespace

----- Original Message -----

**Comment by ironcladlou** [comment]
What about on the other side- `cache.Store`? As a consumer of that interface, I have no knowledge of what namespacing mechanism the reflector might be using to populate the store. All I have is a Get(id) function, an object, and no idea which field to pass to Get or how to make an id from my object that is compatible with whatever scheme the reflector used. I could be misunderstanding the suggested approach.

**Comment by pmorie** [comment]
@ironcladlou Agree, the consumer and reflector might have different ideas about keying. That makes me think there should be distinct interfaces that capture reflector / consumer needs.

**Comment by derekwaynecarr** [comment]
Can a reflector be given a cache func?

**Comment by derekwaynecarr** [comment]
Similar to what we do for generic.EtcdRegistry?

**Comment by smarterclayton** [comment]
The consumer of a reflector should be the person who initialized it.  They know what the uniqueness criteria is.

**Comment by smarterclayton** [comment]
That transformation was what I was thinking of

**Comment by ironcladlou** [comment]
Still confused- the reflectors aren't consumed. cache.Stores are consumed. The reflector is an orthogonal/external actor which populates a cache.Store.

**Comment by smarterclayton** [comment]
Then both of them need to agree what constitutes a unique identifier to Get.

----- Original Message -----

**Comment by smarterclayton** [comment]
For instance, what if you are trying to reflect contents from two deployments at once?

----- Original Message -----

**Comment by ironcladlou** [comment]
@smarterclayton 
>Then both of them need to agree what constitutes a unique identifier to Get.

No argument there, and I checked out the generic.Etcd implementation- what this does imply though is a change to the `cache.Store` methods to accept not just an id, but either a runtime.Object or an id/namespace pair. Agree?

**Comment by smarterclayton** [comment]
An id/namespace pair is somewhat opinionated - it means the cache store only supports 2 means of input.  If your "keyFunc" takes and returns an interface{} then you can pass whatever you want to the store.

----- Original Message -----

**Comment by ironcladlou** [comment]
@smarterclayton 
>An id/namespace pair is somewhat opinionated - it means the cache store only supports 2 means of input.  If your "keyFunc" takes and returns an interface{} then you can pass whatever you want to the store.

Agree. I think we're on the same page now.

**Comment by pmorie** [comment]
@smarterclayton @ironcladlou also agree, key function seems like the best way.  Thinking through the interface changes:

    type Store interface {
        Add(obj interface{})
        Update(obj interface{})
        Delete(obj interface{})
        List() []interface{}
        Get(obj interface{}) (item interface{}, exists bool)

        Replace(objs []interface)

        ContainedIds() util.StringSet
    }

One thing I'm not sure about is whether there should be a distinct `ContainedIds` call now that there's not an idea of `id` for an object in the rest of the interface.  It implies the existence of a key function.

Thoughts?

**Comment by ironcladlou** [comment]
@pmorie 
That's the interface I have been prototyping. As far as `ContainedIds`, I don't see the purpose anymore. In the contexts where `ContainedIds` is used, the IDs are fed back to Get/Update/Delete calls, etc. You could replace those with List, feeding the resulting interface{} instances back to Get/Update/Delete, etc.

ContainedIds in cache.Store is already implemented with a linear iteration over the items map, so it wouldn't even be a performance change in that regard to move the iteration to the caller.

**Comment by ironcladlou** [comment]
@pmorie One implication I haven't yet seen addressed: if most calls to Store now implicitly execute a pluggable key function, and the key function can fail, any Store function which does keying should probably return an `error`. Not pretty, but I'm not sure it's safe to omit the error?

**Comment by ironcladlou** [comment]
Maybe I spoke too soon. I'm still puzzling through how to reproduce [this algorithm](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/client/cache/poller.go#L78-L89) without exposing the key function from the store (for example).

**Comment by ironcladlou** [comment]
>Maybe I spoke too soon. I'm still puzzling through how to reproduce this algorithm without exposing the key function from the store (for example).

Now I'm not sure why Poller.sync is doing individual Updates with a final Delete of all not-Updated items. Could it be implemented instead by collecting all the new items in a local []interface{} and calling Store.Replace atomically? That sidesteps the requirement to have a means to perform interface{} equality checks via KeyFunc.

**Comment by ironcladlou** [comment]
I've opened a PR which a POC implementation of namespacing cache support:

https://github.com/GoogleCloudPlatform/kubernetes/pull/3810

Discussion welcome. I'll hold off continuing to work on it until I get some feedback.

**Comment by erictune** [comment]
I think Poller can be killed.
The only use is in  plugin/pkg/scheduler/factory/factory.go
and there is a TODO there to switch to NewReflector, which I think can now be done.

**Comment by pmorie** [comment]
Even better @erictune

**Comment by erictune** [comment]
This is serious because it prevents kubelets from seeing two pods with the same name on GCE.

**Comment by erictune** [comment]
I just wrote a test which fails and shouldn't.

**Comment by erictune** [comment]
@ironcladlou @pmorie @deads2k is one of you actively working on a fix?

**Comment by smarterclayton** [comment]
We use poller downstream.  We need status on nodes and pods to be watchable before we can remove it.

**Comment by ironcladlou** [comment]
@erictune 
>@ironcladlou @pmorie @deads2k is one of you actively working on a fix?

You figured this out already, but for the sake of visibility, let it be known that I am working the issue via #3810.

P.S., we are currently using Poller downstream because of our inability to receive watch events in response to Pod container state changes (which aren't persisted in etcd). We'd really love to ditch Poller, but can't without the watch functionality.

**Comment by ironcladlou** [comment]
This issue should be closed by https://github.com/GoogleCloudPlatform/kubernetes/pull/3810

**Comment by goltermann** [comment]
#3810 is merged.
